### ISSUE TITLE
Custom Telemetry not Logging

### ISSUE THREAD

**Issue by yarajpal** [microsoft/ApplicationInsights-dotnet#1484]
Hi,

I am using application insights for logging to Geneva using Microsoft.ApplicationInsights.AspNetCore nuget package.

1. Used Dependency Injection to configure Telemetry Configuration.
2. Used Geneva Agent Channel.

I can see the logs for Application Insights SDKs incoming HTTP requests(Request Telemetry) , but I can't see my custom trace logs within the incoming request.

The strange thing is I can see all the custom logs when running the app locally. The issue is only when deploying on Azure cloud on Service Fabric Cluster.

I am using telemetryClient.TrackTrace() for custom logs.

Suggest anything I am missing here?
I am using Microsoft.ApplicationInsights.AspNetCore" Version="2.8.2",
.Net Framework 4.7.2

Thanks
Yash

**Comment by lmolkova** [comment]
Please post your configuration code  here: both how you initialize via DI and also how you initialize `TelemetryClient` for custom telemetry.

Repro app should be also very helpful. 

Since you are not seeing custom logs from deployed app only, you might need to check
-  your MA config on the service and see if you properly configured OneDsProviders
- instrumentation key you supply to channel on the cluster and if it matches one in the MA config

**Comment by yarajpal** [comment]
```
public void ConfigureServices(IServiceCollection services)
 {
            services.AddMvc();
            services.AddApplicationInsightsTelemetry();
            services.AddSingleton<ITelemetryChannel>(new GenevaAgentChannel(((IGenevaTelemetryConfiguration)FrontDoorServiceConfiguration).InstrumentationKey));
            services.AddSingleton<ITelemetryInitializer>(new TelemetryInitializer(
                new Dictionary<string, string>
                {
                    { MonitoringConstants.Environment, ((IFeatureManagementConfiguration)FrontDoorServiceConfiguration).EnvironmentTier },
                    { MonitoringConstants.Region, ((IFeatureManagementConfiguration)FrontDoorServiceConfiguration).ClusterLocation },
                    { MonitoringConstants.Cluster, ((IFeatureManagementConfiguration)FrontDoorServiceConfiguration).ClusterFriendlyName },
                }));
            services.AddSingleton<ITelemetryInitializer, FabricTelemetryInitializer>();
            services.AddMvc();
 }

```
For custom telemetry I have created one TelemetryClient to be used in entire code as follows
TelemetryClient telemetryClient = new TelemetryClient(configuration);
(The configuration is the same the config from the DI).

```
public static void InitializeTelemetryConfiguration(TelemetryConfiguration telemetryConfiguration, IGenevaTelemetryConfiguration genevaTelemetryConfiguration)
        {
            if (telemetryConfiguration == null) throw new ArgumentException(nameof(telemetryConfiguration));
            if (genevaTelemetryConfiguration == null) throw new ArgumentException(nameof(genevaTelemetryConfiguration));

            TelemetryConfiguration = telemetryConfiguration;

            TelemetryConfiguration.InstrumentationKey = genevaTelemetryConfiguration.InstrumentationKey;
            TelemetryConfiguration.TelemetryChannel.DeveloperMode = false;

            TelemetryConfiguration.TelemetryProcessorChainBuilder.Use(next => new GenevaMetricsProcessor(next)
            {
                MetricAccountName = genevaTelemetryConfiguration.MetricAccountName,
                MetricNamespace = genevaTelemetryConfiguration.MetricNamespace,
                DefaultDimensions =
                {
                    MonitoringConstants.RoleInstance,
                    MonitoringConstants.NodeName,
                    MonitoringConstants.RoleName,
                    MonitoringConstants.Environment,
                    MonitoringConstants.Region,
                    MonitoringConstants.Cluster,
                },
            });

            TelemetryConfiguration.DefaultTelemetrySink.Initialize(telemetryConfiguration);
            TelemetryConfiguration.TelemetryProcessorChainBuilder.Build();
        }
    }

```

- I checked my MA config , its configured correctly.
- I verified the instrumentation key too.
- Also if I am not using dependency injection , then I can see the custom logs. The document says recommended way is using Dependency Injection.

**Comment by lmolkova** [comment]
What is the intention behind `public static void InitializeTelemetryConfiguration(TelemetryConfiguration telemetryConfiguration, IGenevaTelemetryConfiguration genevaTelemetryConfiguration)`  this method?

Where do you get your `genevaTelemetryConfiguration` from ?
Why do you need to copy certain properties from it?
How do you create telemetry client - do you use `TelemetryConfiguration` from this method?
Basically, why do you need another instance of configuration?

I'd recommend to simply use DI to resolve TelemteryClient from the DI container and skip all the steps done in `InitializeTelemetryConfiguration` along with creation of an extra configuration instance

**Comment by yarajpal** [comment]
- The intention behind using the InitializeTelemetryConfiguration method is to have a common method to set the values for Instrumentation key, Metric account name and namespace. We are using this common method for our worker service too.
-  genevaTelemetryConfiguration just gives us the values for MetricAccountName, MetricNamespace ,and Instrumentationkey.
- The TelemetryConfiguration that we pass here is the one we take from Di container.
```
public static void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (app == null) throw new ArgumentNullException(nameof(app));

            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseHealthChecks(HealthCheckEndpoint);
            app.UseAuthentication();
            app.UseMvc();

            GenevaInitializer.InitializeTelemetryConfiguration(
                app.ApplicationServices.GetService<TelemetryConfiguration>(),
                FrontDoorConfiguration.Instance);
        }

```
Also I am using the same instance of Configuration from the DI.

**Comment by lmolkova** [comment]
Ok, so you do not create another instance of TelemetryConfiguration. Good.
How do you initialize TelemteryClient (that one that you use to TrackTrace)?


It feels like `InitializeTelemetryConfiguration ` method should be reduced to GenevaMetricsProcessor configuration. Things like `TelemetryConfiguration.InstrumentationKey = genevaTelemetryConfiguration.InstrumentationKey` are questionable: if by chance ikey you used to configure the channel is different, it may lead to some issues. So in the best case, it's redundant.

And calling this: `TelemetryConfiguration.DefaultTelemetrySink.Initialize(telemetryConfiguration);` is again in the best case, redundant.

Assuming you DO have different Ikeys in `new GenevaAgentChannel(((IGenevaTelemetryConfiguration)FrontDoorServiceConfiguration).InstrumentationKey)` and `genevaTelemetryConfiguration.InstrumentationKey` you might get some discrepancies like the one that you are reporting.

Can you do something like this:
``` 
            string ikey = ((IGenevaTelemetryConfiguration)FrontDoorServiceConfiguration).InstrumentationKey;
            var channel = new GenevaAgentChannel(ikey)
            services.AddApplicationInsightsTelemetry(ikey);
            services.AddSingleton<ITelemetryChannel>(() => channel);
```

and reduce the `InitializeTelemetryConfiguration` to something like

```
            var metricProcessor = new GenevaMetricsProcessor(next)
            {
                MetricAccountName = genevaTelemetryConfiguration.MetricAccountName,
                MetricNamespace = genevaTelemetryConfiguration.MetricNamespace,
                DefaultDimensions =
                {
                    MonitoringConstants.RoleInstance,
                    MonitoringConstants.NodeName,
                    MonitoringConstants.RoleName,
                    MonitoringConstants.Environment,
                    MonitoringConstants.Region,
                    MonitoringConstants.Cluster,
                };
            };
            // this is new and important!
            metricProcessor.Initialize(TelemetryConfiguration);

            telemetryConfiguration.TelemetryProcessorChainBuilder.Use(next => metricProcessor);

            telemetryConfiguration.TelemetryProcessorChainBuilder.Build();
```

**Comment by lmolkova** [comment]
Here is the test app configured in a similar way to send auto-collected ASP.NET Core telemetry to Geneva + custom trace telemetry.  [trace telemetry and requests.zip](https://github.com/microsoft/ApplicationInsights-dotnet/files/3949083/trace.telemetry.and.reuqests.zip). Hope it helps.
